### PR TITLE
feat(stdlib): Implement List.zip, List.unzip, List.zipWith, Array.reduceRight, Array.zipWith

### PR DESCRIPTION
**Pull request by alexsnezhko3**
- Added zip and unzip to list to bring it up to parity with array
- Added reduceRight to array
- Added zipWith to both array and list

### CODE REVIEW

**Review by ospencer — APPROVED** [review]
This is awesome, thank you! Docs look really good too. 🎉 

(for anyone like me doing mental gymnastics about whether `zipWith` should be called `map2`, we've already got `Option.zipWith`.)

**Comment by ospencer** [comment]
I thought I fixed the snapshots to not change on standard library changes like this, but maybe not? @alexsnezhko3 Can you run `npm run compiler test -- -u` and commit the results?

**Review by marcusroberts — APPROVED** [review]
Excellent contribution thank you!

File: stdlib/list.gr [hunk 1]
```diff
@@ -548,6 +548,77 @@ export let unique = list => {
...
+ * @example List.zip([1, 2, 3], [4, 5]) // [(1, 4), (2, 5)]
+ * 
+ * @since v0.5.3
+ */
+export let rec zip = (list1, list2) => {
+  match ((list1, list2)) {
+    ([], []) => [],
+    ([first1, ...rest1], [first2, ...rest2]) => [(first1, first2), ...zip(rest1, rest2)],
+    _ => []
+  }
```
Comment (phated) [inline stdlib/list.gr:573]: I think this needs to use a helper function to be made tail-recursive. I believe you can look at `List.length` or `List.reverse` to see the pattern for doing that.

File: stdlib/list.gr [hunk 2]
```diff
@@ -548,6 +548,77 @@ export let unique = list => {
...
+ * @example List.zipWith((a, b) => a * b, [1, 2, 3], [4, 5]) // [4, 10]
+ * 
+ * @since v0.5.3
+ */
+export let rec zipWith = (fn, list1, list2) => {
+  match ((list1, list2)) {
+    ([], []) => [],
+    ([first1, ...rest1], [first2, ...rest2]) => [fn(first1, first2), ...zipWith(fn, rest1, rest2)],
+    _ => []
+  }
```
Comment (phated) [inline stdlib/list.gr:601]: I think this needs to use a helper function to be made tail-recursive. I believe you can look at `List.length` or `List.reverse` to see the pattern for doing that.

**Review by phated — CHANGES_REQUESTED** [review]
Just a note about `List.zip` and `List.zipWith` not being tail-recursive (also, correct me if I'm mistaken).

File: stdlib/list.gr (hunk 1 above)
Comment (alexsnezhko3) [reply stdlib/list.gr:573]: Would something like this work?
```
let zip = (list1, list2) => {
  let rec zipInner = (list1, list2, acc) => {
    match ((list1, list2)) {
      ([first1, ...rest1], [first2, ...rest2]) => zipInner(rest1, rest2, [(first1, first2), ...acc]),
      _ => reverse(acc)
    }
  }
  zipInner(list1, list2, [])
}
```
I add the next elements in the list to the beginning of the accumulator and then reverse it at the end to bring it back to its normal shape

**Review by ospencer — APPROVED** [review]
Implementation looks great!

**Review by phated — APPROVED** [review]
🔥 Awesome! Thanks!

@ospencer Before merging, can you comment on https://github.com/grain-lang/grain/issues/730 about the decision for the List zips functions to operate on lists of different lengths? We also need a followup issue to make the breaking change for Array.zip to match this behavior. 🙏

**Comment by ospencer** [comment]
The array issue is https://github.com/grain-lang/grain/issues/1395 👍

**Pull request merged**